🐳 Add docker-compose

murgle2 opened the pull request.

Hona commented:
Why is this `env-vars.list → .env.txt` instead of just `.env`?

Hona commented on scripts/populate_db_with_fake_data.js:
```diff
@@ -2,7 +2,7 @@
 *   to help with development. Use the following command to run this script:
 *
-*   node <path-to-file>/populate_db_with_fake_data.js
+*   docker exec website-api-1 node ../scripts/populate_db_with_fake_data.js
```
Use `docker-compose exec api node ....`

Hona commented on docker-compose.override.yml:
```diff
@@ -0,0 +1,32 @@
+version: '3.8'
+
+services:
+  db:
+    platform: linux/x86_64
+    volumes:
+      - ./scripts/init-db:/docker-entrypoint-initdb.d
+    environment:
+      MYSQL_DATABASE: 'momentum'
+      MYSQL_USER: 'mom'
+      MYSQL_PASSWORD: 'password'
+      MYSQL_ROOT_PASSWORD: 'root'
+
+  api:
+    platform: linux/x86_64
+    volumes:
+      - ./scripts:/app/scripts
+      - ./server/config:/app/server/config
+      - ./server/docs:/app/server/docs
+      - ./server/public:/app/server/public
+      - ./server/src:/app/server/src
+      - ./server/test:/app/server/test
+    build:
+      context: ./server
+      dockerfile: Dockerfile
+    entrypoint: sh -c "npm install -g nodemon --unsafe-perm && nodemon server.js"
+
+  integration-tests:
+    extends: api
+    entrypoint: sh -c "npm test"
+    environment:
+      NODE_ENV: 'test'
```
I'd name this file something more meaningful: what is `override`?

murgle2 replied:
I'm not opposed to naming it docker-compose.dev.yml but that would require running `docker-compose -f docker-compose.yml -f docker-compose.dev.yml up -d` instead of just `docker-compose up -d` when developing. I'd prefer using override because of this [https://docs.docker.com/compose/extends/#example-use-case](https://docs.docker.com/compose/extends/#example-use-case)

Hona commented on .env.txt:
```diff
@@ -1,21 +1,22 @@
+# Required for dev, the Steam API web key (you can get it from https://steamcommunity.com/dev/apikey)
```
I'd rename this file to .env.TEMPLATE or something like that

murgle2 commented:
> Why is this `env-vars.list → .env.txt` instead of just `.env`?

I added description in readme but pretty much the .env.txt is copied to same directory and renamed .env. This .env is part of gitignore so all env vars will be kept locally.

Hona commented on docker-compose.prod.yml:
```diff
@@ -0,0 +1,17 @@
+  db:
+    restart: always
+    environment:
+      MYSQL_DATABASE: 'momentum'
+      MYSQL_USER: ${MOM_DATABASE_USER}
+      MYSQL_PASSWORD: ${MOM_DATABASE_PW}
+      MYSQL_ROOT_PASSWORD: ${MYSQL_ROOT_PW}
```
For all these docker files, we set the DB connection info here - why not set this as env vars for the API to fetch? That way it isn't needed in the .env?

murgle2 replied:
Not sure what you mean here, can you elaborate more?

Hona commented on docker-compose.yml:
```diff
@@ -0,0 +1,16 @@
+    env_file:
+      - .env
+    ports:
+      - ${NODE_PORT}:${NODE_PORT}
```
Here we should always expose the same port internally, ie:

In the Dockerfile
```
EXPOSE 3200
```

docker-compose
```
${NODE_PORT}:3200
```

Hona commented on server/.dockerignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
+npm-debug.log
```
Add .env or any other secrets - we don't want your secrets accidentally inside of images - you should still be able to mount or use .env files later - it just won't send it to the build context now

Hona commented on server/README.md:
````diff
@@ -4,75 +4,45 @@
+The initial database content will need to be created (tables, constraints, etc.). Run the force sync DB script to do so once the database container is initialized:
 ```
-node <path-to-file>/force_sync_db.js
+docker exec website-api-1 node ../scripts/force_sync_db.js
````
`docker-compose exec api etc..`

Hona commented on server/README.md:
````diff
@@ -4,75 +4,45 @@
+Restart the api and you are good to go:
 ```
-nodemon server.js
+docker restart website-api-1
````
`docker-compose up -d --force-recreate api`

Hona commented on server/README.md:
```diff
@@ -4,75 +4,45 @@
+Navigate to `http://localhost:3002/`. The app will automatically reload if you change any of the source files.
 
+### Applying Database Updates
+If updates or additions are made to the database design after you create the database tables for the first time, you will most likely have to apply these updates to your local databases. To do this, simply run the force_sync_db.js script again to clear and recreate the initial database content.
```
> simply run the `docker-compose exec ....` again

Agree?

Hona commented on server/README.md:
````diff
@@ -4,75 +4,45 @@
+Tests will start automatically when you run docker compose up -d from above. To re-run the tests, restart the testing container:
 ```
-NODE_ENV=test
+docker restart website-integration-tests-1
````
`docker-compose up -f docker-compose.tests.yml -d --force-recreate integration-tests`

murgle2 replied:
Currently the test db and dev db run on the same container and the integration-tests service starts in the dev env as its in the same compose file.
If we move tests to a separate docker file, do you think we should also move the test db to this docker file where it can run in its own container?

murgle2 replied:
Using another compose file for test env

Hona commented on server/README.md:
```diff
@@ -4,75 +4,45 @@
-### Set up Testing Database
-In your MySQL command line, run:
+## Prod Setup
+Update the .env file and set this variable:
```
> Copy the `.env.TEMPLATE` file to `.env` and set this variable:

TheNerdMan commented:
Pulled this down, tried running it but its not working. 

Please update the README to include comands/files that need to be filled out for development.

![image](https://user-images.githubusercontent.com/17830051/141167484-fe28b73e-ace6-4264-b2f9-d269bd69fa86.png)
_The above screenshot was taken after copying env.TEMPLATE to .env and entering my SteamWebAPI key and SQL users

TheNerdMan commented:
Using the correct version of docker-compose I've notice 2 things.

1.  This is not using the new typescript development. I know its a WIP at but if we can start from using that then all new development will be on the right page.
2. It's completely missing the front end client. If this is to be used as a dev environment it needs both.

murgle2 commented:
Note: this relies on the new tests outlined in Use docker-compose integration tests (#557)

Hona commented on scripts/force_sync_db.js:
```diff
@@ -2,7 +2,7 @@
 *   to run this script:
 *
-*   node <path-to-file>/force_sync_db.js
+*   docker compose exec api node ../scripts/force_sync_db.js
```
Typo? `docker-compose`

Hona commented on scripts/populate_db_with_fake_data.js:
```diff
@@ -2,7 +2,7 @@
 *   to help with development. Use the following command to run this script:
 *
-*   node <path-to-file>/populate_db_with_fake_data.js
+*   docker compose exec api node ../scripts/populate_db_with_fake_data.js
```
Typo? docker-compose

Hona commented on server/README.md:
````diff
@@ -4,75 +4,50 @@
+The initial database content will need to be created (tables, constraints, etc.). Run the force sync DB script to do so once the database container is initialized:
 ```
-nodemon server.js
+docker compose exec api node ../scripts/force_sync_db.js
````
Typo? docker-compose

Hona commented on server/README.md:
```diff
@@ -4,75 +4,50 @@
+Navigate to `http://localhost:3002/`. The app will automatically reload if you change any of the source files.
 
+### Applying Database Updates
+If updates or additions are made to the database design after you create the database tables for the first time, you will most likely have to apply these updates to your local databases. To do this, simply run docker-compose exec api node ../scripts/force_sync_db.js again to clear and recreate the initial database content.
```
Typo? docker-compose

Hona commented on server/README.md:
```diff
@@ -4,75 +4,50 @@
-CREATE DATABASE momentum_test;
-CREATE USER mom_test@localhost;
-GRANT ALL PRIVILEGES ON momentum_test.* TO mom_test@localhost;
+docker compose -f docker-compose.yml -f docker-compose.test.yml up -d
```
Typo? docker-compose

Hona commented on server/README.md:
````diff
@@ -4,75 +4,50 @@
+``` 
+To re-run the tests, restart the testing container:
+```
+docker compose up -d --force-recreate api
````
Typo? docker-compose

Hona commented on server/README.md:
````diff
@@ -4,75 +4,50 @@
+From the website directory, run:
 ```
-npm test
+docker compose -f docker-compose.yml -f docker-compose.prod.yml up -d
````
Typo? docker-compose
